azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -1053,4 +1056,63 @@ public void shutdown() {
       this.scheduler.shutdownNow();
     }
   }
+
+  private class PollingCriteria {
+
+    private final Props azkabanProps;
+    private final SystemMemoryInfo memInfo = SERVICE_PROVIDER.getInstance(SystemMemoryInfo.class);
+    private final OsCpuUtil cpuUtil = SERVICE_PROVIDER.getInstance(OsCpuUtil.class);
+
+
+    public boolean shouldPoll() {
+      if (satisfiesFlowThreadsAvailableCriteria() && satisfiesFreeMemoryCriteria()
+          && satisfiesCpuUtilizationCriteria()) {
+        return true;
+      }
+      return false;
+    }
+
+    public PollingCriteria(final Props azkabanProps) {
+      this.azkabanProps = azkabanProps;
+    }
+
+    private boolean satisfiesFlowThreadsAvailableCriteria() {
+      final int flowThreadsAvailableConfig = this.azkabanProps.
+          getInt(ConfigurationKeys.AZKABAN_POLLING_CRITERIA_FLOW_THREADS_AVAILABLE, 0);
+
+      // allow polling if not present or configured with invalid value
+      if (flowThreadsAvailableConfig <= 0) {
+        return true;
+      }
+      final int remainingFlowThreads = FlowRunnerManager.this.getMaxNumRunningFlows() -
+          FlowRunnerManager.this.getNumRunningFlows();
+      return remainingFlowThreads > 0;
+    }
+
+    private boolean satisfiesFreeMemoryCriteria() {
+      final int minFreeMemoryConfigGb = this.azkabanProps.
+          getInt(ConfigurationKeys.AZKABAN_POLLING_CRITERIA_MIN_FREE_MEMORY_GB, 0);
+
+      // allow polling if not present or configured with invalid value
+      if (minFreeMemoryConfigGb <= 0) {
+        return true;
+      }
+      final int minFreeMemoryConfigKb = minFreeMemoryConfigGb * 1024 * 1024;
+      return this.memInfo.isFreePhysicalMemoryAbove(minFreeMemoryConfigKb);
+    }
+
+    private boolean satisfiesCpuUtilizationCriteria() {
+      final double maxCpuUtilizationConfig = this.azkabanProps.
+          getDouble(ConfigurationKeys.AZKABAN_POLLING_CRITERIA_MAX_CPU_UTILIZATION_PCT, 100);
+
+      // allow polling if criteria not present or configured with invalid value
+      if (maxCpuUtilizationConfig <= 0 || maxCpuUtilizationConfig >= 100) {
+        return true;
+      }
+
+      final double cpuLoad = this.cpuUtil.getCpuLoad();
+      return (cpuLoad == -1) ? true : cpuLoad < maxCpuUtilizationConfig;
```

Comment:
Consider using `Double.compare()` instead of `==` to compare double values?

Reply:
Here -1 is an expected exceptional value to indicate a calculation failure. I think it's safe to use == here.